pump/enum_file_writer.go

```diff
@@ -0,0 +1,71 @@
+package pump
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+
+	"github.com/ipfs/go-cid"
+)
+
+type FailedBlocksWriter interface {
+	Write(c cid.Cid) (int, error)
+	Flush() error
+	Count() uint
+}
+
+var _ FailedBlocksWriter = &FileEnumeratorWriter{}
+var _ FailedBlocksWriter = &NullableFileEnumeratorWriter{}
+
+type FileEnumeratorWriter struct {
+	file  *bufio.Writer
+	count uint
+}
+
+type NullableFileEnumeratorWriter struct {
+	count uint
+}
+
+func NewFileEnumeratorWriter(path string) (enumWriter FailedBlocksWriter, close func() error, err error) {
+	fo, err := os.OpenFile(path, os.O_TRUNC|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	w := bufio.NewWriter(fo)
+
+	return &FileEnumeratorWriter{
+		file:  w,
+		count: 0,
+	}, fo.Close, nil
```

Comment:
what about implementing `io.Close()` ?  Then you wouldn't have to pass around this closer function. It'd be more go like.

Reply:
That's definitely a valid alternative. I am currently surfing the wave where you `return` the Close() function so the caller will not forget to call the `io.Close()` as it's easy to make such a mistake or sometimes you don't even think calling the `Close()` is needed. This way it's very explicit. like `context.WithCancel`

Reply:
A big difference imho is that with `context.WithCancel` you are expected to call cancel within the same function as where you created the context. Here, closing can happen in a very different place so you have to wire and pass the closer somehow. If generalized, this pattern could make things very messy.

Reply:
It's just a different pattern. It's normal to pass around the cancel function for context too. I can change it if you prefer the `io.Close()` pattern. Your tool, your rules haha.